oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/ApdexThresholdConfig.java

```diff
@@ -0,0 +1,102 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.oap.server.core.analysis;
+
+import java.io.FileNotFoundException;
+import java.io.Reader;
+import java.io.StringReader;
+import java.util.Collections;
+import java.util.Map;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.skywalking.oap.server.configuration.api.ConfigChangeWatcher;
+import org.apache.skywalking.oap.server.core.Const;
+import org.apache.skywalking.oap.server.core.CoreModule;
+import org.apache.skywalking.oap.server.core.CoreModuleProvider;
+import org.apache.skywalking.oap.server.library.util.ResourceUtils;
+import org.yaml.snakeyaml.Yaml;
+
+/**
+ * Apdex threshold configuration dictionary adapter.
+ * Looking up a service apdex threshold from dynamic config service.
+ *
+ * @author hongtaogao
+ */
+@Slf4j
+public class ApdexThresholdConfig extends ConfigChangeWatcher implements ConfigurationDictionary {
+
+    private static final String CONFIG_FILE_NAME = "service-apdex-threshold.yml";
+
+    private static final int SYSTEM_RESERVED_THRESHOLD = 500;
+
+    private Map<String, Integer> dictionary = Collections.emptyMap();
+
+    private String rawConfig = Const.EMPTY_STRING;
+
+    public ApdexThresholdConfig(final CoreModuleProvider provider) {
+        super(CoreModule.NAME, provider, "apdexThreshold");
+        try {
+            updateConfig(ResourceUtils.read(CONFIG_FILE_NAME));
+        } catch (final FileNotFoundException e) {
+            log.error("Cannot config from: {}", CONFIG_FILE_NAME, e);
+        }
+    }
+
+    @Override public Number lookup(String name) {
+        int t = dictionary.getOrDefault(name, -1);
+        if (t < 0) {
+            t = dictionary.getOrDefault("default", -1);
+        }
+        if (t < 0) {
+            log.warn("Pick up system reserved threshold {}ms because of config missing", SYSTEM_RESERVED_THRESHOLD);
+            return SYSTEM_RESERVED_THRESHOLD;
+        }
+        if (log.isDebugEnabled()) {
+            log.debug("Apdex threshold of {} is {}ms", name, t);
+        }
+        return t;
+    }
+
+    @Override public void notify(ConfigChangeEvent value) {
+        if (EventType.DELETE.equals(value.getEventType())) {
+            activeSetting("");
+        } else {
+            activeSetting(value.getNewValue());
+        }
+    }
+
+    @Override public String value() {
+        return rawConfig;
+    }
+
+    private synchronized void activeSetting(String config) {
+        if (log.isDebugEnabled()) {
+            log.debug("Updating using new static config: {}", config);
+        }
+        rawConfig = config;
+        updateConfig(new StringReader(config));
+    }
+
+    @SuppressWarnings("unchecked")
+    private void updateConfig(final Reader contentRender) {
+        dictionary = (Map<String, Integer>)new Yaml().load(contentRender);
```

Comment:
Missing volatile for the dictionary, as you are creating a new one for it, and the notification is in another thread.

Reply:
It will be invoked by two functions, the first one is constructor and another is under the protection of `synchronize`.

Reply:
Oh, I see. You are using this tricky one.

Reply:
At the first time, I add `synchronize` to `updateConfig`, but `rawCofig` also needs it, then move it to its invoker.